Initial implementation of link identity on chain attestation (WIP). (#589)
radicle-dev/radicle-registry: docs/external-attestations.md

# External Attestations

## Feature goals

The Radicle Registry state contains data about users, organizations and software projects.
However, the same entities will also have an identity off chain, in other systems.

> For instance, a Registry user could also have an identity on Radicle Link (hereinafter "Link"), or on Github, Gitlab, Twitter, or they will have an email address...

Reviewer: This is probably not too early to make this design decision: can a user be linked to multiple identities of the same kind? For example can I have 2 GitHub accounts (1 professional and 1 private) and 3 email addresses (1 from company A, 1 from B and 1 personal)? Should I be able to have multiple link identities?

Author: This is a very good point.

The design of the "external identities collection", where "external" means "outside of Radicle Link and Registry" is probably entirely up to Upstream, and I would not see a problem allowing several email addresses or something similar.
That data would be part of the user profile, there could be anything there.

About Registry and Link identities, of course a user could create more than one
IMHO this would be a bit against the spirit of the platform but there is nothing we can do to prevent it.

However we are not required to make it _easy_ to work with multiple Link and Radicle identities, and this feature should stick to one-to-one attestation of Radice and Link IDs.

Unless the product team thinks this is a scenario we need to actively support...

Reviewer 2: @radicle-dev/product

> For instance, a Registry user could also have an identity on Radicle Link (hereinafter "Link"), or on Github, Gitlab, Twitter, or they will have an email address...

Reviewer 2: ```suggestion
For instance, a Registry user could also have an identity on Radicle Link (hereinafter "Link"), or on Github, Gitlab, Twitter, or email address.
```

In the same way, a registered project could also exist on Link, Github, or elsewhere.

> When relevant it is desirable to be able to link these identities together so that a registry use can be confident that they do in fact represent the same identity.

Reviewer 2: ```suggestion
When relevant it is desirable to be able to link these identities together so that a user of the registry can be confident that they do in fact represent the same identity.
```

It is important to distinguish two different kinds of attestation:

1. Simple linking of identities ("_identity attestation_")
2. Detailed certification of a specific identity revision ("_identity revision attestation_" for platforms that support it, like Link)

The second builds on the first but is an optional extension to it.

## General principle of operation

The feature builds on the ability to refer to one identity from another one.

For instance, it should be possible to insert a reference to a Registry user identity inside a Link user identity (by writing the Registry user ID inside the Link user metadata).
It should also be possible to do the reverse: insert the Link user ID (their URN) inside the Registry user state.

> The key point is that only the legitimate user can do each of these operations.

Reviewer 2: ```suggestion
The key point is that only a legitimate user can do each of these operations.
```

The same concept can be applied to any kind of identity (projects, or even organizations if desired.

## Further details

There are still a few open points.

### Security model

One of them is the "security model" of this attestation: what it proves when everything works as intended is clear, but what to deduce is anything goes wrong is not (like, how to handle user key theft or compromise, or how to show partial attestations where only one of the two links is in place).

Related to the security model is the problem of handling key rotation on both sides of the link (both on the Registry and Link).

### Registry storage cost

Then there is the "registry storage cost" problem: in principle the registry state should contain links to every supported external identity (Link, Github, Twitter...) but this list can be quite long.

An easy solution to this problem is to only store a single link inside the registry, unambiguously referring to an external piece of metadata that in turn will contain all other external links.

> To guarantee absence of ambiguity this link should be a cryptographic hash of the external metadata.

Reviewer: One thing to keep in mind is that every update of the hash is stored on the blockchain too. The data savings here are much smaller than what it seems on the first sight, but simplicity and accessibility decrease significantly.

Author: True, but this growth problem would exist for whatever mutable content we would store.

I mean, if what we store is a hash, the blockchain history will grow containing all the registered hashes, but if we store a full metadata blob it would grow containing all the blobs.

The growth is there anyway, and the space vs data accessibility tradeoff is the same.

One of the reasons why I keep insisting that we should not require users to register every revision of their Link metadata is to mitigate this "growth problem": only revisions containing info worth of being attested should be stored, others can be skipped.

A practical approach to this would be storing the Link identity on the Registry, and putting every other identity references inside the Link identity metadata, reducing the registry state requirements to a single link, which is acceptable.
Since Link identities are versioned with cryptographic hashes the system would be unambiguous, as required.

### Data format and verification

Let's say we decide to store the Link user identity in the Registry user state; we should then decide how to store it.

> In Link identities are represented by Radicle URNs, as described in the (identity resolution)[https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/identity_resolution.md] and (identities)[https://github.com/radicle-dev/radicle-link/pull/248] specifications.

Reviewer 2: ```suggestion
In Link identities are represented by Radicle URNs, as described in the [identity resolution](https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/identity_resolution.md) and [identities](https://github.com/radicle-dev/radicle-link/pull/248) specifications.
```

Author: :facepalm: :ok_hand:

However we have the following issues:

- Registry does not depend on Link.
- The URN format is not compact, while in the registry state every byte counts.

> - The URN format could evolve, or this feature (identity attestation) could evolve, allowing extensions to store something more than Link URNs, or something different; if we wanted Registry to be able to verify the format of this piece of data the core Registry code would need to be updated before each data format change.

Reviewer: It's worth writing down that the data format should be versioned, this CAN NOT be forgotten in the data design process and it MUST be an available feature from day 1

Author: Does line 80 of this document suffice?
It is part of the list of requirements for the MVP.

```
* Specify and implement the compact (and extensible) URN binary encoding format (in Link).
```

I could remark "versioned and extensible" but the intention behind "extensible" is "that can evolve in a formalized way".

Reviewer: Yes, sorry, I haven't noticed it :)
